File: backend/services/otpService.js
const bcrypt = require('bcryptjs');
const { query } = require('../db');

const OTP_TTL_MINUTES = 5;

function generateCode() {
  return String(Math.floor(100000 + Math.random() * 900000)); // 6 digits
}

// Creates a one-time code, stores it hashed, and "sends" it.
async function requestOtp(phone) {
  const code = generateCode();
  const codeHash = await bcrypt.hash(code, 10);
  const expiresAt = new Date(Date.now() + OTP_TTL_MINUTES * 60 * 1000);

  await query(
    'INSERT INTO otp_codes (phone, code_hash, expires_at) VALUES ($1, $2, $3)',
    [phone, codeHash, expiresAt]
  );

  // TODO: integrate a real SMS provider (e.g. Eskiz / Play Mobile for Uzbekistan).
  if ((process.env.OTP_PROVIDER || 'console') === 'console') {
    console.log(`[OTP] ${phone} -> ${code} (valid ${OTP_TTL_MINUTES}m)`);
  }
  return true;
}

// Verifies the most recent unexpired, unconsumed code for a phone number.
async function verifyOtp(phone, code) {
  const { rows } = await query(
    `SELECT id, code_hash FROM otp_codes
      WHERE phone = $1 AND expires_at > NOW() AND consumed_at IS NULL
      ORDER BY created_at DESC
      LIMIT 1`,
    [phone]
  );
  if (rows.length === 0) return false;

  const match = await bcrypt.compare(code, rows[0].code_hash);
  if (!match) return false;

  await query('UPDATE otp_codes SET consumed_at = NOW() WHERE id = $1', [rows[0].id]);
  return true;
}

module.exports = { requestOtp, verifyOtp };

File: split_output/backend/app/services/otp_service.py
"""OTP generation and verification (replaces services/otpService.js).

Codes are stored hashed (bcrypt), never in plaintext. In development the code is
printed to the server console instead of being sent by SMS.
"""
import datetime as dt
import random

import bcrypt
from sqlalchemy import func
from sqlalchemy.orm import Session

from ..config import OTP_PROVIDER
from ..models import OtpCode

OTP_TTL_MINUTES = 5


def _generate_code() -> str:
    return f"{random.randint(100000, 999999)}"  # 6 digits


def request_otp(db: Session, phone: str) -> bool:
    code = _generate_code()
    code_hash = bcrypt.hashpw(code.encode(), bcrypt.gensalt()).decode()
    expires_at = dt.datetime.now(dt.timezone.utc) + dt.timedelta(minutes=OTP_TTL_MINUTES)

    db.add(OtpCode(phone=phone, code_hash=code_hash, expires_at=expires_at))
    db.commit()

    # TODO: integrate a real SMS provider (e.g. Eskiz / Play Mobile for Uzbekistan).
    if (OTP_PROVIDER or "console") == "console":
        print(f"[OTP] {phone} -> {code} (valid {OTP_TTL_MINUTES}m)")
    return True


def verify_otp(db: Session, phone: str, code: str) -> bool:
    record = (
        db.query(OtpCode)
        .filter(
            OtpCode.phone == phone,
            OtpCode.expires_at > func.now(),
            OtpCode.consumed_at.is_(None),
        )
        .order_by(OtpCode.created_at.desc())
        .first()
    )
    if record is None:
        return False
    if not bcrypt.checkpw(code.encode(), record.code_hash.encode()):
        return False

    record.consumed_at = func.now()
    db.commit()
    return True
